inference_overview.md

```diff
@@ -0,0 +1,449 @@
+inferences.
+
+Common bid prediction models use contextual, ad and user features for
+prediction. Out of these, only the user features are private. The machine
+learning models using such private features are the only ones that need to run
+inside the TEE. Contextual and ad features are not private, so pre-computation
+on these can be done outside the TEE.
+
+For example, a model that relies only on contextual features can be run in the
+contextual path during the RTB request. A model based on only ad features can
+pre-compute results per ad and upload them to the K/V server. Such results can
+be queried during request processing and made available to `generateBid()`.
+
+On the other hand, a model that uses user features, or one that uses user and ad
+features will need to run inside the TE, and cannot run outside the TEE
+boundary.
+
+### Unified models
+
+A **unified** approach to such a prediction could run a model that takes
+contextual, ad and user features as input, and makes a prediction, running the
+entire model inside the TEE. B&A will support unified models directly through
+the [two ways][19] of running inference (embedded and external models) described
+above. In the unified approach, the entire model will need to run inside the
+TEE, since it uses private features. These models will tend to be larger. Both
+these factors can increase the cost and latency of running such models.
+
+### Factorized models
+
+**Model factorization** is a technique that makes it possible to break a single
+model into multiple pieces, and then combine those pieces into a prediction. In
+B&A-supported use cases, models often make use of three kinds of data: user
+data, contextual data, and ad data.
+
+As described above, in the unified case, a single model is trained on all three
+kinds of data. In the factorized case, you could break the model up into
+multiple pieces. The model that includes the sensitive user data needs to be
+executed within the trust boundary, on the buyer's bidding services.
+
+This makes the following design possible:
+
+1. Break the model up into a private piece (the sensitive user data) and one or
+  more non-private pieces (the contextual and ad data).
+1. Optionally, inferences from some or all of the non-private pieces can be
+  passed in as arguments to a [User Defined Function][1] (UDF), e.g.
+  `generateBid()`. For example, contextual embeddings can be computed along the
+  contextual path, returned to the seller's frontend in the contextual response,
+  and then can be passed into the B&A auction as part of the `buyer_signals`
+  constructed by the seller when they initiate the Auction.
+1. Optionally, ad techs can create models for non-private pieces (such as ads
+  features) ahead of time, then materialize embeddings from those models into
+  the ad retrieval server and fetch these embeddings at runtime.
+1. To make a prediction during a UDF, combine private embeddings from the
```

Comment:
For clarity:

To generate a prediction within a UDF, combine private embeddings (from a TEE-based inference service) with non-private embeddings (sourced from UDF arguments or the ad retrieval store) using a dot product operation.

Reply:
Done.